src/orange/surf/detail/InvoluteSolver.hh

```diff
@@ -0,0 +1,306 @@
+
+    static CELER_CONSTEXPR_FUNCTION real_type tol()
+    {
+        if constexpr (std::is_same_v<real_type, double>)
+            return 1e-8;
+        else if constexpr (std::is_same_v<real_type, float>)
+            return 1e-5;
+    }
+
+    //// ACCESSORS ////
+
+    //! Get involute parameters
+    CELER_FUNCTION real_type r_b() const { return r_b_; }
+    CELER_FUNCTION real_type a() const { return a_; }
+    CELER_FUNCTION Sign sign() const { return sign_; }
+
+    //! Get bounds of the involute
+    CELER_FUNCTION real_type tmin() const { return tmin_; }
+    CELER_FUNCTION real_type tmax() const { return tmax_; }
+
+  private:
+    //// DATA ////
+    // Involute parameters
+    real_type r_b_;
+    real_type a_;
+    Sign sign_;
+
+    // Bounds
+    real_type tmin_;
+    real_type tmax_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from involute parameters.
+ */
+CELER_FUNCTION InvoluteSolver::InvoluteSolver(
+    real_type r_b, real_type a, Sign sign, real_type tmin, real_type tmax)
+    : r_b_(r_b), a_(a), sign_(sign), tmin_(tmin), tmax_(tmax)
+{
+    CELER_EXPECT(r_b > 0);
+    CELER_EXPECT(a >= 0);
+    CELER_EXPECT(tmax > 0);
+    CELER_EXPECT(tmin >= 0);
+    CELER_EXPECT(tmax < 2 * constants::pi + tmin);
+}
+
+//---------------------------------------------------------------------------//
+
+/*!
+ * Find all roots for involute surfaces that are within the bounds and result
+ * in positive distances. Performed by doing a Regular Falsi Iteration on the
+ * root function, \f[
+ * f(t) = r_b * [v{cos(a+t) + tsin(a+t)} + u{sin(a+t) - tcos(a+t)}] + xv - yu
+ * \f]
+ * where the Regular Falsi Iteration is given by: \f[
+ * t_c = [t_a*f(t_b) - t_b*f(t_a)] / [f(t_b) - f(t_a)]
+ * \f]
+ * where \em t_c replaces the bound with the same sign (e.g. \em t_a and \em
+ * t_b
+ * ). The initial bounds can be determined by the set of: \f[
+ * {0, beta - a, beta - a - pi, beta - a + pi, beta - a - 2pi, beta - a + 2pi
+ * ...} /f] Where \em beta is: \f[ beta = arctan(-v/u) \f]
+ */
+CELER_FUNCTION auto
+InvoluteSolver::operator()(Real3 const& pos,
+                           Real3 const& dir,
+                           SurfaceState on_surface) const -> Intersections
+{
+    using constants::pi;
+    // Flatten pos and dir in xyz and uv respectively
+    real_type x = pos[0];
+    real_type const y = pos[1];
+
+    real_type u = dir[0];
+    real_type v = dir[1];
+
+    // Mirror systemm for counterclockwise involutes
+    if (sign_)
+    {
+        x = -x;
+        u = -u;
+    }
+
+    /*
+     * Define tolerances.
+     * tol_point gives the tolerance level for a point,
+     * account for the floating point error when performing square roots.
+     * tol_conv gives the tolerance for the Regular Falsi iteration.
+     */
+    real_type tol_conv = r_b_ * tol();
```

Comment:
Move these tolerances closer to where they're being used. You can just inline the `tol_conv` into the solver definition.